### PR TITLE
chore(vpa): update dependencies

### PR DESCRIPTION
**Pull request by davidspek**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This PR updates the dependency package versions for the vertical pod autoscaler. It also updates the `update-codegen.sh` script that needed fixing after https://github.com/kubernetes/autoscaler/pull/7572 which depended on the updated packages.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by adrianmoisey** [comment]
/remove-kind feature
/kind cleanup
/ok-to-test

Thanks for the PR, can you also bump the VPA's dependencies, as this only does the e2e tests

**Comment by davidspek** [comment]
@adrianmoisey There are too many changes for GitHub to display them in the PR. But if you look at https://github.com/davidspek/autoscaler/tree/feat/update-vpa-k8s/vertical-pod-autoscaler the dependencies were updated for the regular VPA as well.

**Comment by adrianmoisey** [comment]
So far this looks good to me. I am getting some failures on the e2e tests locally, but I think those aren't related. I'll spend some more time looking into that to make sure it's unrelated.

**Comment by davidspek** [comment]
Great, thanks!

**Comment by adrianmoisey** [comment]
There are two small linting issues:
```
$ golangci-lint run
pkg/recommender/input/cluster_feeder.go:165:25: SA1019: cache.NewIndexerInformer is deprecated: Use NewInformerWithOptions instead. (staticcheck)
	indexer, controller := cache.NewIndexerInformer(
	                       ^
pkg/utils/vpa/api.go:84:25: SA1019: cache.NewIndexerInformer is deprecated: Use NewInformerWithOptions instead. (staticcheck)
	indexer, controller := cache.NewIndexerInformer(vpaListWatch,
	                       ^
```

(if you run this locally, you may see more linting issues, but a PR was merged into master to take care of those, the ones I pasted above is the output after I merged master into your branch, locally).

**Comment by adrianmoisey** [comment]
/lgtm

e2e tests passed on my laptop.

Something to note: not all the dependencies have been upgraded. I don't mind doing those in a later PR, unless you want to do them now

**Comment by davidspek** [comment]
I can do them here. Which ones did I miss?

**Comment by davidspek** [comment]
I see I missed `k8s.io/legacy-cloud-providers`. Just pushed a commit to update that as well.

**Comment by davidspek** [comment]
Oh I thought I'd scope this to just the k8s dependencies. Should I change it to all?

**Comment by adrianmoisey** [comment]
> Oh I thought I'd scope this to just the k8s dependencies. Should I change it to all?

Oh I see, I didn't realise that.

I think it's worth bumping them all at once

**Comment by davidspek** [comment]
Pushed a commit upgrading the other dependencies.

**Comment by davidspek** [comment]
@adrianmoisey Can you give this another `lgtm`?

**Comment by adrianmoisey** [comment]
/lgtm 

For whomever approves...
Do these dependencies relate to the compatibility chart in the readme: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/docs/installation.md#compatibility

File: vertical-pod-autoscaler/e2e/go.mod [hunk 1]
```diff
@@ -3,214 +3,199 @@
...
+	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.31.3
+	k8s.io/endpointslice => k8s.io/endpointslice v0.31.3
+	k8s.io/kms => k8s.io/kms v0.31.3
+	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.31.3
+	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.31.3
+	k8s.io/kube-proxy => k8s.io/kube-proxy v0.31.3
+	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.31.3
+	k8s.io/kubectl => k8s.io/kubectl v0.31.3
+	k8s.io/kubelet => k8s.io/kubelet v0.31.3
+	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.31.3
```
Comment (tobschli) [inline vertical-pod-autoscaler/e2e/go.mod:194]:
```suggestion
```

This module does not have this tag, as it was removed from the staging dir of kubernetes ([ref](https://github.com/kubernetes/kubernetes/pull/124767))
The same suggestion applies to the `go.mod` of the e2e directory.

Comment (davidspek) [reply vertical-pod-autoscaler/e2e/go.mod]: Thanks for the catch. Removed.

File: vertical-pod-autoscaler/e2e/v1/e2e.go [hunk 2]
```diff
@@ -236,7 +236,7 @@ func setupSuite() {
 	// #41007. To avoid those pods preventing the whole test runs (and just
 	// wasting the whole run), we allow for some not-ready pods (with the
 	// number equal to the number of allowed not-ready nodes).
-	if err := e2epod.WaitForPodsRunningReady(context.TODO(), c, metav1.NamespaceSystem, int32(framework.TestContext.MinStartupPods), int32(framework.TestContext.AllowedNotReadyNodes), podStartupTimeout); err != nil {
```
Comment (raywainman) [inline vertical-pod-autoscaler/e2e/v1/e2e.go:239]: Looks like we are losing the `framework.TestContext.AllowedNotReadyNodes` here, I haven't looked yet but do we know why? Is it OK?

**Comment by raywainman** [comment]
/lgtm

The delta in changes here is crazy! Over 500k less lines of code!

2 super quick things:

1. Left a comment about the modification to the test, just want to quickly validate that this is intended.
2. Can we change the PR title to specify that this is for the `e2e` tests? Makes things easier when scanning changes for release notes.

**Comment by raywainman** [comment]
> For whomever approves...
> Do these dependencies relate to the compatibility chart in the readme: 
> https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/docs/installation.md#compatibility

My understanding is that this is partially affecting this - it means our tests with this VPA version will likely only work in the later Kubernetes versions. We'll have to be more careful about updating the compatibility table when we make the same changes in the main VPA code.

File: vertical-pod-autoscaler/e2e/v1/e2e.go (hunk 2 above)
Comment (raywainman) [reply vertical-pod-autoscaler/e2e/v1/e2e.go:239]: Found the original PR that removed this: https://github.com/kubernetes/kubernetes/pull/124205

So looks generally harmless? Would love to validate that `framework.TestContext.AllowedNotReadyNodes` was always 0 in our case as well.

Comment (adrianmoisey) [reply vertical-pod-autoscaler/e2e/v1/e2e.go:239]: For what it's worth, I've run the e2e tests locally and everything passes

**Comment by adrianmoisey** [comment]
> 2\. Can we change the PR title to specify that this is for the `e2e` tests? Makes things easier when scanning changes for release notes.

GitHub breaks on this change, it's both the VPA's dependencies and the e2e tests's dependencies.

> > For whomever approves...
> > Do these dependencies relate to the compatibility chart in the readme:
> > https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/docs/installation.md#compatibility
> 
> My understanding is that this is partially affecting this - it means our tests with this VPA version will likely only work in the later Kubernetes versions. We'll have to be more careful about updating the compatibility table when we make the same changes in the main VPA code.

I've been testing it locally using [Kubernetes 1.26.3](https://github.com/kubernetes/autoscaler/blob/16210e3da03def7e7cd5cc55012dd2ff6ffd8683/vertical-pod-autoscaler/hack/run-e2e-locally.sh#L77) and it works.

What I don't know is what the promises of client-go are regarding backwards compatibility, and how we determine when to update that compatibility table in the VPA's docs.

**Comment by davidspek** [comment]
@raywainman This PR also updates the dependencies for the main VPA, not just `e2e`. The PR is just to large to show it on GitHub. In terms of the change to the e2e test, those were necessary after the upgrade.

In terms of general compatibility, the cluster-autoscaler is already on `v0.33.0-alpha.0`. I didn't do that here and went to the latest released version instead. However, for a project like VPA I think it's very important to stay up to date with Kubernetes dependencies and test against the latest versions, similar to what the cluster-autoscaler does.

**Comment by ialidzhikov** [comment]
Wrt to all the comments that the diff is huge. We should remove the `vendor/`. I suggest to do it in another PR. The cluster-autoscaler seems to already remove the `vendor/` dir.

**Comment by adrianmoisey** [comment]
What about:
1. Close this PR
2. Make a new PR that removes `vendor/`
3. Get that PR merged
4. Make a new PR to bump dependencies 

It may make review much easier

**Comment by ialidzhikov** [comment]
I created issue for the vendor dir removal: https://github.com/kubernetes/autoscaler/issues/7570

**Comment by davidspek** [comment]
Once https://github.com/kubernetes/autoscaler/pull/7572 is merged I'll update this PR. There's a fix necessary after https://github.com/kubernetes/autoscaler/pull/7572 which relies on the dependencies being updated so I'd like to get that in quickly after this change. I'll create a new PR for the `e2e` dependency updates once https://github.com/kubernetes/autoscaler/pull/7573 is merged.

File: vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml
Comment (adrianmoisey) [inline vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:1]: Is it worth removing this, and only scoping the PR to dependencies only?

Comment (davidspek) [reply vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:1]: Sure

**Pull request merged**